# Ensemble Learning

In [14]:
## Initial Imports
import warnings
warnings.filterwarnings('ignore')

In [15]:
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter

In [16]:
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

In [143]:
## Read the CSV and Perform Basic Data Cleaning

# Load the data
df = pd.read_csv("../Data/LoanStats_2019Q1.csv")
# see al the columns
df.columns


Index(['loan_amnt', 'int_rate', 'installment', 'home_ownership', 'annual_inc',
       'verification_status', 'issue_d', 'loan_status', 'pymnt_plan', 'dti',
       'delinq_2yrs', 'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_bal',
       'total_acc', 'initial_list_status', 'out_prncp', 'out_prncp_inv',
       'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_amnt', 'next_pymnt_d', 'collections_12_mths_ex_med',
       'policy_code', 'application_type', 'acc_now_delinq', 'tot_coll_amt',
       'tot_cur_bal', 'open_acc_6m', 'open_act_il', 'open_il_12m',
       'open_il_24m', 'mths_since_rcnt_il', 'total_bal_il', 'il_util',
       'open_rv_12m', 'open_rv_24m', 'max_bal_bc', 'all_util',
       'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_si

In [144]:
df.dtypes

loan_amnt                     float64
int_rate                      float64
installment                   float64
home_ownership                 object
annual_inc                    float64
                               ...   
total_bal_ex_mort             float64
total_bc_limit                float64
total_il_high_credit_limit    float64
hardship_flag                  object
debt_settlement_flag           object
Length: 86, dtype: object

In [145]:
df.columns.to_series().groupby(df.dtypes).groups

{float64: ['loan_amnt', 'int_rate', 'installment', 'annual_inc', 'dti', 'delinq_2yrs', 'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_bal', 'total_acc', 'out_prncp', 'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int', 'total_rec_late_fee', 'recoveries', 'collection_recovery_fee', 'last_pymnt_amnt', 'collections_12_mths_ex_med', 'policy_code', 'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m', 'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il', 'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc', 'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m', 'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util', 'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct', 'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl', 'mort_acc', 'mths_since_recent_bc', 'mths_since_recent_inq', 'num_accts_ever_120_pd', 'num_actv_bc_tl', 'num_actv_rev_tl', 'num_bc_sats', 'num_b

In [146]:
# Drop the null columns where all values are null
df = df.dropna(axis='columns', how='all')

# Drop the null rows
df = df.dropna()

In [147]:
df['loan_status'].unique()

array(['low_risk', 'high_risk'], dtype=object)

In [148]:
# class note 11.2
df=pd.get_dummies(df, columns=[
    'home_ownership', 
    'verification_status', 
    'issue_d', 
    'pymnt_plan', 
    'initial_list_status', 
    'next_pymnt_d', 
    'application_type', 
    'hardship_flag', 
    'debt_settlement_flag']
)

In [149]:
# Preview the data
df.head()

loan_amnt  int_rate  installment  annual_inc loan_status    dti  \
0    10500.0    0.1719       375.35     66000.0    low_risk  27.24   
1    25000.0    0.2000       929.09    105000.0    low_risk  20.23   
2    20000.0    0.2000       529.88     56000.0    low_risk  24.26   
3    10000.0    0.1640       353.55     92000.0    low_risk  31.44   
4    22000.0    0.1474       520.39     52000.0    low_risk  18.76   

   delinq_2yrs  inq_last_6mths  open_acc  pub_rec  ...  issue_d_Mar-2019  \
0          0.0             0.0       8.0      0.0  ...                 1   
1          0.0             0.0      17.0      1.0  ...                 1   
2          0.0             0.0       8.0      0.0  ...                 1   
3          0.0             1.0      10.0      1.0  ...                 1   
4          0.0             1.0      14.0      0.0  ...                 1   

   pymnt_plan_n  initial_list_status_f  initial_list_status_w  \
0             1                      0                      1   
1             1                      0                      1   
2             1                      0                      1   
3             1                      0                      1   
4             1                      0                      1   

   next_pymnt_d_Apr-2019  next_pymnt_d_May-2019  application_type_Individual  \
0                      0                      1                            1   
1                      0                      1                            1   
2                      0                      1                            1   
3                      0                      1                            1   
4                      0                      1                            1   

   application_type_Joint App  hardship_flag_N  debt_settlement_flag_N  
0                           0                1                       1  
1                           0                1                       1  
2                           0                1                       1  
3                           0                1                       1  
4                           0                1                       1  

[5 rows x 96 columns]

In [150]:
## Split the Data into Training and Testing

# Create our features
X = df.drop(columns="loan_status")

# Create our target
y = df["loan_status"]

In [151]:
X.describe()

loan_amnt      int_rate   installment    annual_inc           dti  \
count  68817.000000  68817.000000  68817.000000  6.881700e+04  68817.000000   
mean   16677.594562      0.127718    480.652863  8.821371e+04     21.778153   
std    10277.348590      0.048130    288.062432  1.155800e+05     20.199244   
min     1000.000000      0.060000     30.890000  4.000000e+01      0.000000   
25%     9000.000000      0.088100    265.730000  5.000000e+04     13.890000   
50%    15000.000000      0.118000    404.560000  7.300000e+04     19.760000   
75%    24000.000000      0.155700    648.100000  1.040000e+05     26.660000   
max    40000.000000      0.308400   1676.230000  8.797500e+06    999.000000   

        delinq_2yrs  inq_last_6mths      open_acc       pub_rec  \
count  68817.000000    68817.000000  68817.000000  68817.000000   
mean       0.217766        0.497697     12.587340      0.126030   
std        0.718367        0.758122      6.022869      0.336797   
min        0.000000        0.000000      2.000000      0.000000   
25%        0.000000        0.000000      8.000000      0.000000   
50%        0.000000        0.000000     11.000000      0.000000   
75%        0.000000        1.000000     16.000000      0.000000   
max       18.000000        5.000000     72.000000      4.000000   

           revol_bal  ...  issue_d_Mar-2019  pymnt_plan_n  \
count   68817.000000  ...      68817.000000       68817.0   
mean    17604.142828  ...          0.177238           1.0   
std     21835.880400  ...          0.381873           0.0   
min         0.000000  ...          0.000000           1.0   
25%      6293.000000  ...          0.000000           1.0   
50%     12068.000000  ...          0.000000           1.0   
75%     21735.000000  ...          0.000000           1.0   
max    587191.000000  ...          1.000000           1.0   

       initial_list_status_f  initial_list_status_w  next_pymnt_d_Apr-2019  \
count           68817.000000           68817.000000           68817.000000   
mean                0.123879               0.876121               0.383161   
std                 0.329446               0.329446               0.486161   
min                 0.000000               0.000000               0.000000   
25%                 0.000000               1.000000               0.000000   
50%                 0.000000               1.000000               0.000000   
75%                 0.000000               1.000000               1.000000   
max                 1.000000               1.000000               1.000000   

       next_pymnt_d_May-2019  application_type_Individual  \
count           68817.000000                 68817.000000   
mean                0.616839                     0.860340   
std                 0.486161                     0.346637   
min                 0.000000                     0.000000   
25%                 0.000000                     1.000000   
50%                 1.000000                     1.000000   
75%                 1.000000                     1.000000   
max                 1.000000                     1.000000   

       application_type_Joint App  hardship_flag_N  debt_settlement_flag_N  
count                68817.000000          68817.0                 68817.0  
mean                     0.139660              1.0                     1.0  
std                      0.346637              0.0                     0.0  
min                      0.000000              1.0                     1.0  
25%                      0.000000              1.0                     1.0  
50%                      0.000000              1.0                     1.0  
75%                      0.000000              1.0                     1.0  
max                      1.000000              1.0                     1.0  

[8 rows x 95 columns]

In [152]:
# Check the balance of our target values
y.value_counts()

low_risk     68470
high_risk      347
Name: loan_status, dtype: int64

In [153]:
# Split the X and y into X_train, X_test, y_train, y_test
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1, stratify=y)
X_train.shape

(51612, 95)

## Data Pre-Processing

Scale the training and testing data using the `StandardScaler` from `sklearn`. Remember that when scaling the data, you only scale the features data (`X_train` and `X_testing`).

In [154]:
# Create the StandardScaler instance
from sklearn.preprocessing import StandardScaler
data_scaler = StandardScaler()

In [155]:
# Fit the Standard Scaler with the training data
# When fitting scaling functions, only train on the training dataset
data_scaler.fit(X_train)

StandardScaler()

In [156]:
# Scale the training and testing data
X_train = data_scaler.transform(X_train)
X_test = data_scaler.transform(X_test)

## Ensemble Learners

In this section, you will compare two ensemble algorithms to determine which algorithm results in the best performance. You will train a Balanced Random Forest Classifier and an Easy Ensemble classifier . For each algorithm, be sure to complete the folliowing steps:

1. Train the model using the training data. 
2. Calculate the balanced accuracy score from sklearn.metrics.
3. Display the confusion matrix from sklearn.metrics.
4. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.
5. For the Balanced Random Forest Classifier only, print the feature importance sorted in descending order (most important feature to least important) along with the feature score

Note: Use a random state of 1 for each algorithm to ensure consistency between tests

### Balanced Random Forest Classifier

In [157]:
# Resample the training data with the BalancedRandomForestClassifier
# 11.3
from imblearn.ensemble import BalancedRandomForestClassifier
brf = BalancedRandomForestClassifier(n_estimators=100, random_state=1)
brf.fit(X_train, y_train)

BalancedRandomForestClassifier(random_state=1)

In [159]:
# Calculated the balanced accuracy score
y_pred_rf = brf.predict(X_test)
balanced_accuracy_score(y_test, y_pred_rf)

0.7871246640962729

In [160]:
# Display the confusion matrix
confusion_matrix(y_test, y_pred_rf)

array([[   58,    29],
       [ 1582, 15536]])

In [161]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred_rf))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.04      0.67      0.91      0.07      0.78      0.59        87
   low_risk       1.00      0.91      0.67      0.95      0.78      0.62     17118

avg / total       0.99      0.91      0.67      0.95      0.78      0.62     17205



In [162]:
# List the features sorted in descending order by feature importance
indices = np.argsort(brf.feature_importances_)[::-1]
for i in indices:
    print(X.columns[i])

total_rec_prncp
total_rec_int
total_pymnt_inv
total_pymnt
last_pymnt_amnt
int_rate
out_prncp
dti
max_bal_bc
issue_d_Jan-2019
installment
annual_inc
bc_util
mths_since_recent_inq
total_bc_limit
out_prncp_inv
avg_cur_bal
issue_d_Mar-2019
revol_bal
mo_sin_old_rev_tl_op
total_bal_ex_mort
tot_cur_bal
mo_sin_old_il_acct
total_bal_il
bc_open_to_buy
il_util
tot_hi_cred_lim
all_util
total_il_high_credit_limit
num_rev_accts
total_rev_hi_lim
mths_since_recent_bc
pct_tl_nvr_dlq
total_acc
mths_since_rcnt_il
mo_sin_rcnt_rev_tl_op
num_rev_tl_bal_gt_0
loan_amnt
inq_last_12m
num_bc_tl
num_op_rev_tl
mo_sin_rcnt_tl
num_actv_bc_tl
total_cu_tl
percent_bc_gt_75
num_sats
num_actv_rev_tl
open_acc
acc_open_past_24mths
inq_fi
num_il_tl
open_il_24m
num_bc_sats
open_act_il
open_rv_24m
mort_acc
total_rec_late_fee
num_tl_op_past_12m
open_acc_6m
next_pymnt_d_Apr-2019
open_rv_12m
open_il_12m
issue_d_Feb-2019
next_pymnt_d_May-2019
inq_last_6mths
delinq_2yrs
num_accts_ever_120_pd
tot_coll_amt
home_ownership_OWN
home_ow

### Easy Ensemble Classifier

In [163]:
# Train the Classifier
from imblearn.ensemble import EasyEnsembleClassifier
eec = EasyEnsembleClassifier(n_estimators=100, random_state=1)
eec.fit(X_train, y_train)

EasyEnsembleClassifier(n_estimators=100, random_state=1)

In [164]:
# Calculated the balanced accuracy score
y_pred_eec = eec.predict(X_test)
balanced_accuracy_score(y_test, y_pred_eec) 

0.9254565671948463

In [165]:
# Display the confusion matrix
confusion_matrix(y_test, y_pred_eec)

array([[   79,     8],
       [  978, 16140]])

In [166]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred_rf))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.04      0.67      0.91      0.07      0.78      0.59        87
   low_risk       1.00      0.91      0.67      0.95      0.78      0.62     17118

avg / total       0.99      0.91      0.67      0.95      0.78      0.62     17205



### Final Questions

1. Which model had the best balanced accuracy score?

    EasyEnsembleClassifier has a better balanced accuracy (0.925).
    

2. Which model had the best recall score?

    Both have the same recall score (0.91 to 0.77) 
    

3. Which model had the best geometric mean score?

    Both have the same geometric mean score (0.78)
    

4. What are the top three features?

    From the Balanced Random Forest Classifier, the top features are 
    total_rec_prncp, 
    total_rec_int, 
    total_pymnt_inv .
